bee-trace/bee-trace/src/subscriber/layer/log.rs

```diff
@@ -0,0 +1,273 @@
+/// is initialised.
+pub struct LogLayer {
+    make_writers: Vec<LogTargetMakeWriter>,
+    fmt_events: LogFormatter,
+}
+
+impl<S> Layer<S> for LogLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        // If the event is originally issued by the `log` crate, generate the appropriate `tracing` metadata.
+        if let Some(metadata) = event.normalized_metadata() {
+            let mut buf = String::new();
+
+            for make_writer in &self.make_writers {
+                // Only write to an output if the event target is enabled by filters.
+                if make_writer.enabled(&metadata, &ctx) {
+                    let mut writer = make_writer.make_writer();
+
+                    if self.fmt_events.format_event(&mut buf, &writer, event).is_ok() {
+                        let _ = io::Write::write(&mut writer, buf.as_bytes());
+                    }
+
+                    buf.clear();
+                }
+            }
+        }
+    }
+}
+
+impl LogLayer {
+    /// The name that specifies the standard output as a log target (instead of a file).
+    const STDOUT_NAME: &'static str = "stdout";
+
+    pub(crate) fn new(config: LoggerConfig) -> Result<Self, Error> {
+        let fmt_events = LogFormatter {
+            color_enabled: config.color_enabled(),
+            target_width: config.target_width(),
+            level_width: config.level_width(),
+        };
+
+        let make_writers = config
+            .outputs()
+            .iter()
+            .map(|output_config: &LoggerOutputConfig| {
+                let level = output_config.level_filter().as_trace();
+
+                let mut targets = if output_config.target_filters().is_empty() {
+                    filter::Targets::default().with_default(level)
+                } else {
+                    let mut targets = filter::Targets::default().with_default(LevelFilter::OFF);
+
+                    for filter in output_config.target_filters() {
+                        targets = targets.with_target(filter.clone().to_lowercase(), level);
+                    }
+
+                    targets
+                };
+
+                for exclusion in output_config.target_exclusions() {
+                    targets = targets.with_target(exclusion.clone().to_lowercase(), LevelFilter::OFF);
+                }
+
+                let dest = match output_config.name().as_str() {
+                    Self::STDOUT_NAME => LogDest::Stdout,
+                    name => {
+                        let file = OpenOptions::new().write(true).create(true).append(true).open(name)?;
+                        LogDest::File(Mutex::new(file))
+                    }
+                };
+
+                Ok(LogTargetMakeWriter::new(LogTarget { filter: targets, dest }))
+            })
+            .collect::<Result<_, _>>()
```

Comment:
Can you add the `io::Error` to the collect's turbofish so you don't have to do it in the closure?

Reply:
Didn't notice this -- yeah that would be better.